src/bsoncxx/document/element.cpp

```diff
@@ -32,11 +34,11 @@
     iter.next_off = _offset; \
     bson_iter_next(&iter)
 
-#define BSONCXX_TYPE_CHECK(name)                            \
-    do {                                                    \
-        if (type() != bsoncxx::type::name) {                \
-            throw std::runtime_error("type is not " #name); \
-        }                                                   \
+#define BSONCXX_TYPE_CHECK(name)                                                            \
+    do {                                                                                    \
+        if (type() != bsoncxx::type::name) {                                                \
+            throw bsoncxx::exception{make_error_code(bsoncxx_error::k_wrong_element_type)}; \
```

Comment:
I've changed this to spell out different error types for the various value types instead.